### PR TITLE
feat: Implemented a new rest endpoints for the concor contrbution and contr files 

### PR DESCRIPTION
**Pull request by atharmajeed**
…ibution files.

## What

[Link to story](https://dsdmoj.atlassian.net/browse/AP-XXX)

DCES-155, DCES-183 - Implemented new rest endpoints for the concor contribution and contribution files.

## Checklist

Before you ask people to review this PR:

- [ ] Tests should be passing: `./gradlew test`
- [ ] Github should not be reporting conflicts; you should have recently run `git rebase main`.
- [ ] Avoid mixing whitespace changes with code changes in the same commit. These make diffs harder to read and conflicts more likely.
- [ ] You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- [ ] You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            log.info("Updating the concor contribution file ref  -> {}", contributionRequest);
+            contributionFileEntity = contributionFileMapper.toContributionFileEntity(contributionRequest);
+            // when file is not provided then xml generate.
+            contributionFileEntity.setFileName(Optional.ofNullable(contributionRequest.getXmlFileName()).orElse(getFilename()));
+            contributionFileEntity.setUserCreated(DB_USER_NAME);
+            contributionFileEntity.setUserModified(DB_USER_NAME);
+
+            contributionFileRepository.save(contributionFileEntity);
+        } catch (Exception e) {
+            throw new MAATCourtDataException("ConcorContributionRequest to ContributionEntity Mapper error.");
```
Comment (stone-dj) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java:62]: I think this exception message could be misleading, as an exception could be thrown by a failure to persist the entity, while the mapping from `ConcorContributionRequest -> ContributionFilesEntity` was successful.

You should consider including exception message in the `MAATCourtDataException` raised.

e.g.
```
String message = "Failed to map ConcorContributionRequest to ContributionFilesEntity and persist: [%s]".formatted(e.getMessage());
throw new MAATCourtDataException(message);
```

File: maat-court-data-api/src/main/java/gov/uk/courtdata/repository/ConcorContributionsRepository.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import gov.uk.courtdata.dces.enums.ConcorContributionStatus;
+import gov.uk.courtdata.entity.ConcorContributionsEntity;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+import java.util.List;
+
+@Repository
+public interface ConcorContributionsRepository extends JpaRepository<ConcorContributionsEntity, Integer> {
+    List<ConcorContributionsEntity> findByStatus(ConcorContributionStatus status);
+    List<ConcorContributionsEntity> findByIdIn(List<String> ids);
```
Comment (stone-dj) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/repository/ConcorContributionsRepository.java:12]: I think the method signature would communicate the intent of this method better by using a `Set` to hold the `ids` as I would imagine the result of calling `findByIdIn` in all the scenarios below would be identical?

```
- findByIdIn(Arrays.asList("1", "3"));
- findByIdIn(Arrays.asList("1", "3", "1"));
- findByIdIn(Arrays.asList("3", "1"));
- findByIdIn(Arrays.asList("3", "3", "1"));
```

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+    private boolean updateConcorStatusForContribution(final List<String> ids, final ConcorContributionStatus status, final Integer contributionFileId ){
+
+        final List<ConcorContributionsEntity> concorFileList = concorRepository.findByIdIn(ids);
+
+        log.info("Concor Contributions for status update - count {}", concorFileList.size());
+        concorFileList.forEach(cc -> {
+            cc.setStatus(status);
+            cc.setContribFileId(contributionFileId);
+            cc.setUserModified("TOGDATA");
```
Comment (stone-dj) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java:75]: Can you extract `"TOGDATA" `as a constant, like you did for `"MLA"`?

e.g. `DB_USER_NAME = "MLA";`

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        return concorFileList.stream().map(ConcorContributionsEntity::getCurrentXml).toList();
+    }
+
+    @Transactional(rollbackFor = MAATCourtDataException.class)
+    public boolean createContributionAndUpdateConcorStatus(final ConcorContributionRequest contributionRequest){
+
+        ValidationUtils.isNull(contributionRequest,"contributionRequest object is null");
+        ValidationUtils.isEmpty(contributionRequest.getContributionIds(),"ContributionIds are empty/null.");
+
+        final ContributionFilesEntity contributionFilesEntity = createContributionFile(contributionRequest);
```
Comment (stone-dj) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java:45]: It's my understanding that this method `createContributionAndUpdateConcorStatus` will **NOT** roll back both DB changes in the case where the `ContributionFilesEntity` is successfully persisted but the number of `ConcorContributionsEntity` found via ids are **zero**

Can you confirm that this is the expected behavior?

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/controller/ConcorContributionsRestController.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+    @ApiResponse(responseCode = "200", content = @Content(mediaType = MediaType.APPLICATION_JSON_VALUE))
+    @ApiResponse(responseCode = "500", description = "Server Error.", content = @Content(mediaType = MediaType.APPLICATION_JSON_VALUE, schema = @Schema(implementation = ErrorDTO.class)))
+    @ApiResponse(responseCode = "404", description = "Not Found.", content = @Content(mediaType = MediaType.APPLICATION_JSON_VALUE, schema = @Schema(implementation = ErrorDTO.class)))
+    @GetMapping(value = "/concor-contribution-files", produces = MediaType.APPLICATION_JSON_VALUE)
+    @Operation(description = "Get a list of Contributions files")
+    public ResponseEntity<List<String>> getConcorContributionFiles(@RequestParam(name = "status") final ConcorContributionStatus status,
+                                                                   @RequestHeader(value = CourtDataConstants.LAA_TRANSACTION_ID, required = false)
+                                                                   final String laaTransactionId) {
+        MDC.put(LoggingData.LAA_TRANSACTION_ID.getValue(), laaTransactionId);
```
[the remark this answers is not in the document]
Comment (mtac50) [reply maat-court-data-api/src/main/java/gov/uk/courtdata/dces/controller/ConcorContributionsRestController.java:46]: As long as the service calling this endpoint is using micrometer for tracing we can remove the `laaTransactionId`. I'll let @atharmajeed confirm but I think the DCES service does use the right tracing packages.

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/controller/ConcorContributionsRestController.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    @ApiResponse(responseCode = "404", description = "Not Found.", content = @Content(mediaType = MediaType.APPLICATION_JSON_VALUE, schema = @Schema(implementation = ErrorDTO.class)))
+    @PostMapping(value = "/update-contribution", produces = MediaType.APPLICATION_JSON_VALUE)
+    @Operation(description = "update contribution file with concatenated concor files")
+    public ResponseEntity<Boolean> updateContributionFileStatus(@RequestBody @NotEmpty final ConcorContributionRequest request,
+                                                                @RequestHeader(value = CourtDataConstants.LAA_TRANSACTION_ID, required = false)
+                                                                final String laaTransactionId) {
+        MDC.put(LoggingData.LAA_TRANSACTION_ID.getValue(), laaTransactionId);
+        log.info("Update concor contribution file references with request {}", request);
+        boolean response = concorContributionsService.createContributionAndUpdateConcorStatus(request);
```
Comment (tariqhossain) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/dces/controller/ConcorContributionsRestController.java:63]: Based on discussion with Mike and Core team, can we change this end point name to CreateContributionFile

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/controller/ConcorContributionsRestController.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        log.info("findContributionFiles count {}", concorFiles.size());
+        return ResponseEntity.ok(concorFiles);
+    }
+
+    @ApiResponse(responseCode = "200", content = @Content(mediaType = MediaType.APPLICATION_JSON_VALUE))
+    @ApiResponse(responseCode = "500", description = "Server Error.", content = @Content(mediaType = MediaType.APPLICATION_JSON_VALUE, schema = @Schema(implementation = ErrorDTO.class)))
+    @ApiResponse(responseCode = "404", description = "Not Found.", content = @Content(mediaType = MediaType.APPLICATION_JSON_VALUE, schema = @Schema(implementation = ErrorDTO.class)))
+    @PostMapping(value = "/update-contribution", produces = MediaType.APPLICATION_JSON_VALUE)
+    @Operation(description = "update contribution file with concatenated concor files")
+    public ResponseEntity<Boolean> updateContributionFileStatus(@RequestBody @NotEmpty final ConcorContributionRequest request,
```
Comment (tariqhossain) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/dces/controller/ConcorContributionsRestController.java:58]: Based on discussion with Mike and Core team, can we change this end point name to CreateContributionFile

File: maat-court-data-api/src/test/java/gov/uk/courtdata/dces/controller/ConcorContributionsRestControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    @Test
+    void testContributionFileContentWhenQueryParamIsNotProvided() throws Exception {
+        mvc.perform(MockMvcRequestBuilders.get(String.format(ENDPOINT_URL  +"/concor-contribution-files"))
+                        .contentType(MediaType.APPLICATION_JSON_VALUE))
+                .andExpect(status().isBadRequest());
+    }
+
+    @Test
+    void testUpdateContributionFileStatus() throws Exception {
+        final ConcorContributionRequest concorContributionRequest = ConcorContributionRequest.builder()
```
[the remark this answers is not in the document]
Comment (stone-dj) [reply maat-court-data-api/src/test/java/gov/uk/courtdata/dces/controller/ConcorContributionsRestControllerTest.java:75]: In this case `final` does not guarantee any immutability, as we can still call `set` methods on the `concorContributionRequest`.

For immutability you also need to mark all fields in `ConcorContributionRequest` as final and ensure that any object returned is also either immutable or a defensive copy. The set of `contributionIds` which is returned by `getContributionIds` is mutable.

`private Set<String> contributionIds;`

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class ConcorContributionsService {
+    private static final String DB_USER_NAME = "MLA";
+    private final ConcorContributionsRepository concorRepository;
+    private final ContributionFilesRepository contributionFileRepository;
+    private final ContributionFileMapper contributionFileMapper;
+
+    public List<String> getConcorFiles(final ConcorContributionStatus status) {
```
Comment (mtac50) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java:34]:
```suggestion
    public List<String> getConcorFiles(ConcorContributionStatus status) {
```

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        log.info("Getting concor contribution file with status with the -> {}", status);
+        final List<ConcorContributionsEntity> concorFileList = concorRepository.findByStatus(status);
+        return concorFileList.stream().map(ConcorContributionsEntity::getCurrentXml).toList();
+    }
+
+    @Transactional(rollbackFor = MAATCourtDataException.class)
+    public boolean createContributionAndUpdateConcorStatus(final ConcorContributionRequest contributionRequest){
```
Comment (mtac50) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java:41]:
```suggestion
    public boolean createContributionAndUpdateConcorStatus(ConcorContributionRequest contributionRequest){
```

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+        ValidationUtils.isNull(contributionRequest,"contributionRequest object is null");
+        ValidationUtils.isEmpty(contributionRequest.getContributionIds(),"ContributionIds are empty/null.");
+
+        final ContributionFilesEntity contributionFilesEntity = createContributionFile(contributionRequest);
+        return updateConcorStatusForContribution(contributionRequest.getContributionIds(), SENT, contributionFilesEntity.getId());
+    }
+
+    private ContributionFilesEntity createContributionFile(final ConcorContributionRequest contributionRequest) {
```
Comment (mtac50) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java:50]:
```suggestion
    private ContributionFilesEntity createContributionFile(ConcorContributionRequest contributionRequest) {
```

File: maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+            contributionFileEntity.setUserModified(DB_USER_NAME);
+
+            contributionFileRepository.save(contributionFileEntity);
+        } catch (Exception e) {
+            throw new MAATCourtDataException("Failed to map ConcorContributionRequest to ContributionFilesEntity and persist: [%s]".formatted(e.getMessage()));
+        }
+        return contributionFileEntity;
+    }
+
+    private boolean updateConcorStatusForContribution(final Set<String> ids, final ConcorContributionStatus status, final Integer contributionFileId ){
```
Comment (mtac50) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/dces/service/ConcorContributionsService.java:68]:
```suggestion
    private boolean updateConcorStatusForContribution(Set<String> ids, ConcorContributionStatus status, Integer contributionFileId ){
```

File: maat-court-data-api/src/main/java/gov/uk/courtdata/util/ValidationUtils.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
+package gov.uk.courtdata.util;
+
+import gov.uk.courtdata.exception.ValidationException;
+import lombok.experimental.UtilityClass;
+
+import java.util.Collection;
+
+@UtilityClass
+public class ValidationUtils {
+    public static <T> void isEmpty(final Collection<T> collection, final String errorMessage) {
```
Comment (mtac50) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/util/ValidationUtils.java:10]:
```suggestion
    public static <T> void isEmpty(Collection<T> collection, String errorMessage) {
```

File: maat-court-data-api/src/main/java/gov/uk/courtdata/util/ValidationUtils.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+        if (collection == null || collection.isEmpty() || containsAnyNull(collection)) {
+            raise(errorMessage);
+        }
+    }
+
+    public static void isNull(final Object o, final String errorMessage) {
```
Comment (mtac50) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/util/ValidationUtils.java:16]:
```suggestion
    public static void isNull(Object o, String errorMessage) {
```

File: maat-court-data-api/src/main/java/gov/uk/courtdata/util/ValidationUtils.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+        if (o == null) {
+            raise(errorMessage);
+        }
+    }
+
+    private static <T> boolean containsAnyNull(final Collection<T> collection) {
```
Comment (mtac50) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/util/ValidationUtils.java:22]:
```suggestion
    private static <T> boolean containsAnyNull(Collection<T> collection) {
```

File: maat-court-data-api/src/main/java/gov/uk/courtdata/util/ValidationUtils.java [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+        for (final T element : collection) {
+            if (element == null) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static void raise(final String errorMessage) {
```
Comment (mtac50) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/util/ValidationUtils.java:31]:
```suggestion
    private static void raise(String errorMessage) {
```

**Review by stone-dj — APPROVED** [review]
Changes look good

**Pull request merged**